precise/iaf_psc_alpha_ps.cpp

```diff
@@ -361,6 +358,7 @@ nest::iaf_psc_alpha_ps::update( Time const& origin, const long from, const long
       // update membrane potential
       if ( not S_.is_refractory_ )
       {
+        // If we do not use V_.expm1_tau_m_*S_.V_m_ + S_.V_m_ here, the accuracy is a bit worse.
```

Comment:
@heplesser This is a very long answer to your question, but the accuracy compared to a theoretical value is higher when using `exp()`. If we use `S_.V_m_ * std::exp( -V_.h_ms_ / P_.tau_m_ )` instead of `V_.expm1_tau_m_ * S_.V_m_ + S_.V_m_` two tests fail, `testsuite/unittests/test_iaf_ps_dc_t_accuracy.sli` and `testsuite/unittests/test_iaf_ps_dc_accuracy.sli` (the two tests I have changed below). The tests "compares spikes generated to the theoretical value for different computation step sizes and interpolation orders". `testsuite/unittests/test_iaf_ps_dc_t_accuracy.sli` writes out information, and if we use `exp()` instead of `expm1()` we get, for interpolation order 3
```
order: 3, exact value of membrane potential after 5 ms is 4.70003629245736 ms.

               h in ms              alpha_presc [ms]      error           [ms]          delta_canon [ms]        error         [ms]
--------------------------------------------------------------------------------------------------------------------------------
                       0          4.70003629245736      8.88178419700125e-16          4.70003629245736      8.88178419700125e-16  
                      -1          4.70003629245736                         0          4.70003629245736                         0  
                      -2          4.70003629245736                         0          4.70003629245736                         0  
                      -3          4.70003629245735      6.21724893790088e-15          4.70003629245736      8.88178419700125e-16  
                      -4          4.70003629245737      1.33226762955019e-14          4.70003629245735      4.44089209850063e-15  
                      -5          4.70003629245736      1.77635683940025e-15          4.70003629245736      8.88178419700125e-16  
                      -6          4.70003629245733      2.48689957516035e-14          4.70003629245735      1.77635683940025e-15  
                      -7          4.70003629245728      7.72715225139109e-14          4.70003629245736      5.32907051820075e-15  
                      -8          4.70003629245749      1.35891298214119e-13          4.70003629245736      8.88178419700125e-16  
                      -9          4.70003629245741      5.24025267623074e-14          4.70003629245736                         0  
                     -10          4.70003629245698      3.77475828372553e-13          4.70003629245736      5.32907051820075e-15  
                     -11          4.70003629245835       9.9475983006414e-13          4.70003629245736      2.66453525910038e-15  
                     -12          4.70003629245631      1.04805053524615e-12          4.70003629245736      5.32907051820075e-15  
                     -13          4.70003629246496      7.60724816473157e-12          4.70003629245733      2.48689957516035e-14  
                     -14          4.70003629246858      1.12256870465899e-11          4.70003629245738       2.1316282072803e-14 
```

When we use `expm1()` as we do now however, we get
```
order: 3, exact value of membrane potential after 5 ms is 4.70003629245736 ms.

               h in ms              alpha_presc [ms]      error           [ms]          delta_canon [ms]        error         [ms]
--------------------------------------------------------------------------------------------------------------------------------
                       0          4.70003629245736      8.88178419700125e-16          4.70003629245736      8.88178419700125e-16  
                      -1          4.70003629245736      8.88178419700125e-16          4.70003629245736                         0  
                      -2          4.70003629245736      8.88178419700125e-16          4.70003629245736                         0  
                      -3          4.70003629245735      1.77635683940025e-15          4.70003629245736      8.88178419700125e-16  
                      -4          4.70003629245735      2.66453525910038e-15          4.70003629245735      4.44089209850063e-15  
                      -5          4.70003629245735      1.77635683940025e-15          4.70003629245736      8.88178419700125e-16  
                      -6          4.70003629245735      2.66453525910038e-15          4.70003629245735      1.77635683940025e-15  
                      -7          4.70003629245736      6.21724893790088e-15          4.70003629245736      5.32907051820075e-15  
                      -8          4.70003629245736      8.88178419700125e-16          4.70003629245736      8.88178419700125e-16  
                      -9          4.70003629245736                         0          4.70003629245736                         0  
                     -10          4.70003629245736      5.32907051820075e-15          4.70003629245736      5.32907051820075e-15  
                     -11          4.70003629245736       3.5527136788005e-15          4.70003629245736      2.66453525910038e-15  
                     -12          4.70003629245736        7.105427357601e-15          4.70003629245736      5.32907051820075e-15  
                     -13          4.70003629245733      2.39808173319034e-14          4.70003629245733      2.48689957516035e-14  
                     -14          4.70003629245738      2.04281036531029e-14          4.70003629245738       2.1316282072803e-14  
```
So for lower `h` the error compared to a theoretical value is lower when using `expm1()`. I can still use `exp()`, but then I will need to change the tolerance in the two tests to make them pass.

Reply:
@stinebuu Thanks for the details! I think we need to take a look at some aspects of that test. The model names in the table header seem wrong (could we take them directly from the list of models tested instead of hardcoding them?), the `iaf_psc_[alpha,delta]_ps` models have no parameter "interpolation order" any more (so no reason to loop over orders), and `iaf_psc_exp_ps` and `iaf_psc_exp_lossless` are not tested.

This leaves the curious matter that the presumably more accurate numerical expression leads to less accurate results. It would be nice to understand that, but that would be a follow-up issue. For now, let us stick to the  V expm1 + V, which gives best performance. That then also should be used in the function computing V during smaller steps and during interpolation. In the comment, maybe point explicitly to this test.